tests/queries/tests.py

```diff
@@ -3259,6 +3260,19 @@ def test_ticket_21748_complex_filter(self):
         self.assertEqual(str(qs1.query).count('INNER JOIN'),
                          str(qs2.query).count('INNER JOIN'))
 
+    def test_non_bool_deprecation_warning(self):
+        with warnings.catch_warnings(record=True) as warns:
+            warnings.simplefilter('always')
+            ModelA.objects.filter(Q(b__name__isnull=1)).exists()
+
+        self.assertEqual(len(warns), 1)
+        msg = str(warns[0].message)
+        self.assertEqual(
+            msg,
+            'Providing a non-bool value with a isnull lookup is '
+            'untested and undocumented leading to inconsistent behaviour.'
+        )
```

Comment:
Please move tests to the `lookup/tests.py`. Testing a deprecation warning can be simplified, e.g.
```python
    # These test will catch ValueError in Django 4.0 when using non-boolean
    # values with an isnull lookup becomes forbidden.

    def test_isnull_non_boolean_value(self):
        msg = (
            'Using a non-boolean value with an isnull lookup is deprecated, '
            'use True or False instead.'
        )
        with self.assertWarnsMessage(RemovedInDjango40Warning, msg):
            ModelA.objects.filter(b__name__isnull=1).exists()
```
We should add tests for all cases i.e. a normal field, related field, and related field with a multi-column relation (`ForeignObject`).